# 1. The World Bank's international debt data
<p>The overall data overview presents international debt data for various countries, detailing disbursements, interest payments, and principal repayments on external debt, both bilateral and multilateral. It includes indicators such as the amount disbursed and paid on long-term external debt, as well as disbursements and payments from official creditors. Additionally, it highlights interest payments on private nonguaranteed external debt. Each entry in the overview corresponds to a specific country and provides insight into its debt-related transactions, reflecting the complexities and dynamics of global debt management and financing. <a href="https://www.worldbank.org">The World Bank</a> is the organization that provides debt to countries.</p>
<p>We will examine data on foreign debt gathered by The World Bank in this notebook. The dataset includes data on the total debt (in USD) that developing nations owe in a number of different categories. We are going to find the answers to the following questions: </p>
<ul>
<li>What is the total amount of debt that is owed by the countries listed in the dataset?</li>
<li>Which country owns the maximum amount of debt and what does that amount look like?</li>
<li>What is the average amount of debt owed by countries across different debt indicators?</li>
</ul>
<p><img src="World bank.png" /> </p>


In [1]:
#Importing all neccessary libraries

import pandas as pd
import numpy as np
import sqlite3 as sql

<p>The next line of code connects me to my database where the data is located. Firstly, let's <code>SELECT</code> <em>all</em> of the columns from the <code>international_debt</code> table. Also, I'll limit the output to the first ten rows.</p>

In [2]:
#Creating connection to Database file
database = "international_debt.db"
connection = sql.connect(database)

In [3]:
%reload_ext sql

In [4]:
%%sql sqlite:///international_debt.db

SELECT * 
FROM international_debt 
LIMIT 10;

Done.


[('Afghanistan', 'AFG', 'Disbursements on external debt, long-term (DIS, current US$)', 'DT.DIS.DLXF.CD', 72894453.7),
 ('Afghanistan', 'AFG', 'Interest payments on external debt, long-term (INT, current US$)', 'DT.INT.DLXF.CD', 53239440.1),
 ('Afghanistan', 'AFG', 'PPG, bilateral (AMT, current US$)', 'DT.AMT.BLAT.CD', 61739336.9),
 ('Afghanistan', 'AFG', 'PPG, bilateral (DIS, current US$)', 'DT.DIS.BLAT.CD', 49114729.4),
 ('Afghanistan', 'AFG', 'PPG, bilateral (INT, current US$)', 'DT.INT.BLAT.CD', 39903620.1),
 ('Afghanistan', 'AFG', 'PPG, multilateral (AMT, current US$)', 'DT.AMT.MLAT.CD', 39107845),
 ('Afghanistan', 'AFG', 'PPG, multilateral (DIS, current US$)', 'DT.DIS.MLAT.CD', 23779724.3),
 ('Afghanistan', 'AFG', 'PPG, multilateral (INT, current US$)', 'DT.INT.MLAT.CD', 13335820),
 ('Afghanistan', 'AFG', 'PPG, official creditors (AMT, current US$)', 'DT.AMT.OFFT.CD', 100847181.9),
 ('Afghanistan', 'AFG', 'PPG, official creditors (DIS, current US$)', 'DT.DIS.OFFT.CD', 72894453.7)]

# 2. Total amount of debt owed by all the countries
<p>The financial debt of a particular country represents its economic state. But if we were to project this on an overall global scale, how will we approach it?</p>
<p>Let's find out the total amount of debt (in USD) that is owed by the different countries. This will give us a sense of how the overall economy of the entire world is holding up.</p>

In [5]:
%%sql

SELECT country_name, 
    SUM(debt) AS total_debt
FROM international_debt 
GROUP BY country_name;

 * sqlite:///international_debt.db
Done.


[('Afghanistan', 680943227.1),
 ('Albania', 3217067103.8),
 ('Algeria', 522973067.0),
 ('Angola', 71368842500.1),
 ('Armenia', 3834876451.5),
 ('Azerbaijan', 8753732947.199999),
 ('Bangladesh', 35045492840.50001),
 ('Belarus', 23273359790.899998),
 ('Belize', 607062668.8),
 ('Benin', 2535179229.3),
 ('Bhutan', 1692237860.3999999),
 ('Bolivia', 7889146776.8),
 ('Bosnia and Herzegovina', 6685436904.6),
 ('Botswana', 646085255.7),
 ('Brazil', 280623966140.8),
 ('Bulgaria', 6369101653.900001),
 ('Burkina Faso', 2419206260.7),
 ('Burundi', 281989671.59999996),
 ('Cabo Verde', 596636725.4999999),
 ('Cambodia', 4229626566.6),
 ('Cameroon', 86491206346.6),
 ('Central African Republic', 201527131.2),
 ('Chad', 1575430132.7999997),
 ('China', 285793494734.2),
 ('Colombia', 45430117605.2),
 ('Comoros', 57568483.2),
 ('Congo, Dem. Rep.', 2239101510.6),
 ('Congo, Rep.', 1519290572.6999996),
 ('Costa Rica', 8616331219.9),
 ("Cote d'Ivoire", 7141829014.8),
 ('Djibouti', 942100811.9999998),
 ('Dominica', 138044299.79999995),
 ('Dominican Republic', 10351687705.300001),
 ('Ecuador', 17510981502.600002),
 ('Egypt, Arab Rep.', 62077727756.9),
 ('El Salvador', 9411612674.599998),
 ('Eritrea', 132493809.29999998),
 ('Eswatini', 475436690.70000017),
 ('Ethiopia', 13718630481.3),
 ('Fiji', 422780493.5),
 ('Gabon', 3187705162.2),
 ('Gambia, The', 435513543.9),
 ('Georgia', 14379874754.5),
 ('Ghana', 11462311995.200003),
 ('Grenada', 273838275.3),
 ('Guatemala', 7932838900.7),
 ('Guinea', 676612007.9999999),
 ('Guinea-Bissau', 169537430.4),
 ('Guyana', 627922921.9999999),
 ('Haiti', 299424732.9),
 ('Honduras', 2543918981.4999995),
 ('IDA only', 179048127207.30002),
 ('India', 133627060958.40001),
 ('Indonesia', 113435696693.5),
 ('Iran, Islamic Rep.', 463662150.59999996),
 ('Jamaica', 3981335937.9),
 ('Jordan', 8472720622.900001),
 ('Kazakhstan', 70159942693.9),
 ('Kenya', 22622406389.800003),
 ('Kosovo', 948077213.5999999),
 ('Kyrgyz Republic', 4656980236.0),
 ('Lao PDR', 8356392647.200001),
 ('Least developed countries: UN classification', 212880992791.90002),
 ('Lebanon', 29697872618.8),
 ('Lesotho', 435350697.3),
 ('Liberia', 520187064.99999994),
 ('Macedonia, FYR', 3502007528.5),
 ('Madagascar', 1739970527.7000003),
 ('Malawi', 626939901.3000001),
 ('Maldives', 1253465689.1),
 ('Mali', 2604524204.3999996),
 ('Mauritania', 3184696802.1000004),
 ('Mauritius', 4787131190.8),
 ('Mexico', 124596786217.29999),
 ('Moldova', 3338411353.8),
 ('Mongolia', 9847249858.9),
 ('Montenegro', 2992522192.0),
 ('Morocco', 21492166822.4),
 ('Mozambique', 4608729187.5),
 ('Myanmar', 6130613635.299999),
 ('Nepal', 3405713007.2),
 ('Nicaragua', 5399238587.1),
 ('Niger', 2478419343.0000005),
 ('Nigeria', 14149905240.900002),
 ('Pakistan', 45139315398.50001),
 ('Papua New Guinea', 8784109797.2),
 ('Paraguay', 5774787846.8),
 ('Peru', 24091780491.9),
 ('Philippines', 25151706785.000004),
 ('Romania', 42813979498.2),
 ('Russian Federation', 191289057259.2),
 ('Rwanda', 1263036810.8999999),
 ('Samoa', 112946425.5),
 ('Sao Tome and Principe', 44798032.5),
 ('Senegal', 7119017874.0),
 ('Serbia', 14740351568.699999),
 ('Sierra Leone', 417307458.59999996),
 ('Solomon Islands', 105104467.3),
 ('Somalia', 119614074.89999999),
 ('South Africa', 36703940742.5),
 ('South Asia', 247608723990.60004),
 ('Sri Lanka', 26764495009.399998),
 ('St. Lucia', 188052886.50000003),
 ('St. Vincent and the Grenadines', 200061522.59999996),
 ('Sudan', 3835096534.1999993),
 ('Syrian Arab Republic', 1840057923.0),
 ('Tajikistan', 1565019582.1),
 ('Tanzania', 7798052922.500002),
 ('Thailand', 17236647553.9),
 ('Timor-Leste', 71000959.19999999),
 ('Togo', 1021007973.9),
 ('Tonga', 55209304.8),
 ('Tunisia', 14223413457.600002),
 ('Turkey', 151125758035.30002),
 ('Turkmenistan', 95135629.29999998),
 ('Uganda', 6139129245.400001),
 ('Ukraine', 28490304100.300003),
 ('Uzbekistan', 9532699072.7),
 ('Vanuatu', 171626014.5),
 ('Venezuela, RB', 36048260107.7),
 ('Vietnam', 45851299896.09999),
 ('Yemen, Rep.', 

# 3. Numbers of distinct countries
<p>The amount of debt due by <em> Afghanistan</em> can be seen in the first ten rows across the various debt indicators. However, we are unsure of how many distinct nations are included in the table. The country names are repeated because it is common for a nation to have debt in more than one debt indicator. </p>
<p>Extracting the number of unique countries present in the table. </p>

In [6]:
%%sql
SELECT  COUNT(DISTINCT(country_name))
     AS total_distinct_countries
FROM international_debt;

 * sqlite:///international_debt.db
Done.


[(124,)]

# 4. Distinct debt indicators
<p>We can see there are a total of 124 countries present on the table. As we saw in the first section, there is a column called <code>indicator_name</code> that briefly specifies the purpose of taking the debt. However, beside it is another column called <code>indicator_code</code> which symbolizes the category of these debts. Understanding these several debt indicators will enable us to identify potential areas of national debt. </p>

In [7]:
%%sql

SELECT  COUNT(DISTINCT(indicator_code))
     AS distinct_debt_indicators
FROM international_debt

order by distinct_debt_indicators;

 * sqlite:///international_debt.db
Done.


[(25,)]

# 5. Country with the highest debt
<p>"Human beings cannot comprehend very large or very small numbers. It would be useful for us to acknowledge that fact." - <a href="https://en.wikipedia.org/wiki/Daniel_Kahneman">Daniel Kahneman</a>. That is more than <em>3 million <strong>million</strong></em> USD, an amount which is really hard for us to fathom. </p>
<p>Let's now find out the country that owns the highest amount of debt along with the amount. <strong>Note</strong> that this debt is the sum of different debts owed by a country across several categories. This will help to understand more about the country in terms of its socio-economic scenarios.  </p>

In [8]:
%%sql

SELECT country_name, 
    SUM(debt) as total_debt
FROM international_debt
GROUP BY country_name
ORDER BY total_debt desc
Limit 1;

 * sqlite:///international_debt.db
Done.


[('China', 285793494734.2)]

# 6. Average amount of total debt across indicators
<p> <em>China</em> has more debt than other countries. A more in-depth breakdown of China's debts can be found <a href="https://datatopics.worldbank.org/debt/ids/country/CHN">here</a>. </p>
<p>Now that we have a quick rundown of the dataset and some of its summary statistics. The various debt indicators that each country owes are already familiar to us. We can delve further deeper to get the average debt of a nation. This will help us understand how the quantity of debt is distributed throughout the various indicators.</p>

In [9]:
%%sql

SELECT 
    indicator_code AS debt_indicator,
    indicator_name, AVG(debt) as average_debt
    
FROM international_debt
GROUP BY debt_indicator,indicator_name
ORDER BY average_debt desc
Limit 10;

 * sqlite:///international_debt.db
Done.


[('DT.AMT.DLXF.CD', 'Principal repayments on external debt, long-term (AMT, current US$)', 5904868401.499195),
 ('DT.AMT.DPNG.CD', 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', 5161194333.812659),
 ('DT.DIS.DLXF.CD', 'Disbursements on external debt, long-term (DIS, current US$)', 2152041216.8902435),
 ('DT.DIS.OFFT.CD', 'PPG, official creditors (DIS, current US$)', 1958983452.8598356),
 ('DT.AMT.PRVT.CD', 'PPG, private creditors (AMT, current US$)', 1803694101.963265),
 ('DT.INT.DLXF.CD', 'Interest payments on external debt, long-term (INT, current US$)', 1644024067.6508067),
 ('DT.DIS.BLAT.CD', 'PPG, bilateral (DIS, current US$)', 1223139290.39823),
 ('DT.INT.DPNG.CD', 'Interest payments on external debt, private nonguaranteed (PNG) (INT, current US$)', 1220410844.4215188),
 ('DT.AMT.OFFT.CD', 'PPG, official creditors (AMT, current US$)', 1191187963.083064),
 ('DT.AMT.PBND.CD', 'PPG, bonds (AMT, current US$)', 1082623947.6536236)]

# 7. The highest amount of principal repayments
<p>It is obvious that the indicator <code>DT.AMT.DLXF.CD</code> tops the chart of average debt. This category includes repayment of long term debts. Countries take on long-term debt to acquire immediate capital. However, more information about this category can be found <a href="https://datacatalog.worldbank.org/principal-repayments-external-debt-long-term-amt-current-us-0">here</a>. </p>
<p>The fact that the levels of the indicators fluctuate significantly after the second one is an intriguing observation in the aforementioned finding. This suggests that the most serious categories in which the countries owe their debts may be the first two indicators.</p>
<p>Consequently, We can look into this more to determine which nation has the largest debt load in the long-term debt category. (<code>DT.AMT.DLXF.CD</code>).We can gain a more precise understanding of a country's economic situation thanks to this discovery, as not all nations experience the same kind of economic disruptions. </p>

In [10]:
%%sql
SELECT 
    country_name, indicator_name, debt
FROM international_debt
WHERE debt = (SELECT 
                 max(debt)
             FROM international_debt
             WHERE indicator_code='DT.AMT.DLXF.CD');

 * sqlite:///international_debt.db
Done.


[('China', 'Principal repayments on external debt, long-term (AMT, current US$)', 96218620835.7)]

# 8. The most common debt indicator
<p>China has the highest amount of debt in the long-term debt (<code>DT.AMT.DLXF.CD</code>) category. This is verified by <a href="https://data.worldbank.org/indicator/DT.AMT.DLXF.CD?end=2018&most_recent_value_desc=true">The World Bank</a>. Verifying our analyses in this way is usually a good idea because it confirms the accuracy of this study. </p>
<p>As we have seen, the highest category in terms of average debt load is long-term debt. Is it, nonetheless, the most typical measure of the nations' debt? Let's investigate that. </p>

In [11]:
%%sql

SELECT 
    indicator_code,
    COUNT(indicator_code) AS indicator_count
FROM international_debt
GROUP BY indicator_code
ORDER BY indicator_count DESC, indicator_code DESC
LIMIT 15;

 * sqlite:///international_debt.db
Done.


[('DT.INT.OFFT.CD', 124),
 ('DT.INT.MLAT.CD', 124),
 ('DT.INT.DLXF.CD', 124),
 ('DT.AMT.OFFT.CD', 124),
 ('DT.AMT.MLAT.CD', 124),
 ('DT.AMT.DLXF.CD', 124),
 ('DT.DIS.DLXF.CD', 123),
 ('DT.INT.BLAT.CD', 122),
 ('DT.DIS.OFFT.CD', 122),
 ('DT.AMT.BLAT.CD', 122),
 ('DT.DIS.MLAT.CD', 120),
 ('DT.DIS.BLAT.CD', 113),
 ('DT.INT.PRVT.CD', 98),
 ('DT.AMT.PRVT.CD', 98),
 ('DT.INT.PCBK.CD', 84)]

# 9. Other workable debt and conclusion
<p>Every nation included in our dataset has taken out debt in each of the six debt indices. The indicator <code>DT.AMT.DLXF.CD</code> is included on the list as well. This suggests that there is a common economic problem facing all of these nations. However, that is only a portion of the story; it is not the conclusion. </p>
<p>Focusing on the amount of debt, let's determine the highest debt level for each nation's indicator. With this, we will be able to determine what additional likely economic problems a nation may be facing.</p>
<p>We looked at international debt owed by nations in this notebook. From the data, we were able to extract a few summary statistics and discover some fascinating facts and numbers. In order to ensure the accuracy of the investigations, we additionally verified our findings.
</p>

In [12]:
%%sql

SELECT 
    country_name, 
    MAX(debt) AS maximum_debt
FROM international_debt
GROUP BY country_name
ORDER BY maximum_debt DESC
LIMIT 10;

 * sqlite:///international_debt.db
Done.


[('China', 96218620835.7),
 ('Brazil', 90041840304.1),
 ('Russian Federation', 66589761833.5),
 ('Turkey', 51555031005.80001),
 ('South Asia', 48756295898.19999),
 ('Least developed countries: UN classification', 40160766261.6),
 ('IDA only', 34531188113.2),
 ('India', 31923507000.8),
 ('Indonesia', 30916112653.8),
 ('Kazakhstan', 27482093686.4)]